File: day15.pl
#!/usr/bin/env perl 
#===============================================================================
#
#         FILE: day15.pl
#
#        USAGE: ./day15.pl
#
#  DESCRIPTION: Day 15 - Making cookies
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: Mark T. Johnson (mtj), markj@matzsoft.com
# ORGANIZATION: MATZ Software & Consulting
#      VERSION: 1.0
#      CREATED: 12/14/15 22:45:47
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.010;

use List::Util qw( max );


my ( @names, %names );

while ( <DATA> ) {
    chomp;
    s/,//g;

    my @info = split;
    my ( $name, $capacity, $durability, $flavor, $texture, $calories )
        = @info[ 0, 2, 4, 6, 8, 10 ];

    substr $name, -1, 1, q{};
    push @names, $name;
    $names{$name} = {
        capacity   => $capacity,
        durability => $durability,
        flavor     => $flavor,
        texture    => $texture,
        calories   => $calories,
    };
}

my @scores = cookie_scores( 100, @names );

say max( map { $_->[0] } @scores );
say max( map { $_->[0] } grep { $_->[1] == 500 } @scores );



sub cookie_scores {
    my $remaining = shift;
    my $first     = shift;
    my @rest      = @_;

    if ( @rest > 0 ) {
        my @results;

        for my $quantity ( 0 .. $remaining ) {
            $names{$first}{quantity} = $quantity;
            push @results, cookie_scores( $remaining - $quantity, @rest );
        }

        return @results;
    }

    my $score    = 1;
    my $calories = 0;

    $names{$first}{quantity} = $remaining;
    for my $property ( qw( capacity durability flavor texture ) ) {
        my $prop_score = 0;

        for my $name ( @names ) {
            $prop_score += $names{$name}{quantity} * $names{$name}{$property};
        }

        $prop_score = 0 if $prop_score < 0;
        $score *= $prop_score;
    }

    for my $name ( @names ) {
        $calories += $names{$name}{quantity} * $names{$name}{calories};
    }

    return ( [ $score, $calories ] );
}



__END__
Frosting: capacity 4, durability -2, flavor 0, texture 0, calories 5
Candy: capacity 0, durability 5, flavor -1, texture 0, calories 8
Butterscotch: capacity -1, durability 0, flavor 5, texture 0, calories 6
Sugar: capacity 0, durability 0, flavor -2, texture 2, calories 1

File: 2015/day15.swift
//
//         FILE: main.swift
//  DESCRIPTION: day15 - Science for Hungry People
//        NOTES: ---
//       AUTHOR: Mark T. Johnson, markj@matzsoft.com
//    COPYRIGHT: © 2021 MATZ Software & Consulting. All rights reserved.
//      VERSION: 1.0
//      CREATED: 07/08/21 11:45:57
//

import Foundation

// Butterscotch: capacity -1, durability -2, flavor 6, texture 3, calories 8
// Cinnamon: capacity 2, durability 3, flavor -2, texture -1, calories 3
struct Ingredient {
    let name:       String
    let capacity:   Int
    let durability: Int
    let flavor:     Int
    let texture:    Int
    let calories:   Int
    
    init( line: String ) {
        let words = line.split { ": ,".contains( $0 ) }
        
        name       = String( words[0] )
        capacity   = Int( words[2] )!
        durability = Int( words[4] )!
        flavor     = Int( words[6] )!
        texture    = Int( words[8] )!
        calories   = Int( words[10] )!
    }
}

func cookieScores( teaspoons: Int, ingredients: [Ingredient], quantities: [Int] ) -> [ ( Int, Int ) ] {
    if ingredients.count > quantities.count {
        return ( 0 ... teaspoons ).flatMap { quantity -> [ ( Int, Int ) ] in
            let teaspoons = teaspoons - quantity
            let quantities = quantities + [ quantity ]
            return cookieScores( teaspoons: teaspoons, ingredients: ingredients, quantities: quantities )
        }
    }
    
    func sumIt( attributes: [ Int ] ) -> Int {
        let pairs = zip( attributes, quantities )
        return max( 0, pairs.reduce( 0 ) { $0 + $1.0 * $1.1 } )
    }
    
    let capacity   = sumIt( attributes: ingredients.map { $0.capacity } )
    let durability = sumIt( attributes: ingredients.map { $0.durability } )
    let flavor     = sumIt( attributes: ingredients.map { $0.flavor } )
    let texture    = sumIt( attributes: ingredients.map { $0.texture } )
    let calories   = sumIt( attributes: ingredients.map { $0.calories } )

    return [ ( capacity * durability * flavor * texture, calories ) ]
}


func parse( input: AOCinput ) -> [Ingredient] {
    return input.lines.map { Ingredient( line: $0 ) }
}


func part1( input: AOCinput ) -> String {
    let ingredients = parse( input: input )
    let scores = cookieScores( teaspoons: 100, ingredients: ingredients, quantities: [] )
    
    return "\( scores.max( by: { $0.0 < $1.0 } )!.0 )"
}


func part2( input: AOCinput ) -> String {
    let ingredients = parse( input: input )
    let scores = cookieScores( teaspoons: 100, ingredients: ingredients, quantities: [] )
    
    return "\( scores.filter { $0.1 == 500 }.max( by: { $0.0 < $1.0 } )!.0 )"
}


try runTests( part1: part1 )
try runTests( part2: part2 )
try solve( part1: part1 )
try solve( part2: part2 )
